Riot/Modules/Analytics/Analytics.swift

```diff
@@ -0,0 +1,250 @@
+/// and sending events through this client.
+@objcMembers class Analytics: NSObject {
+    
+    // MARK: - Properties
+    
+    /// The singleton instance to be used within the Riot target.
+    static let shared = Analytics()
+    
+    /// The analytics client to send events with.
+    private var client: AnalyticsClientProtocol = PostHogAnalyticsClient()
+    
+    /// Whether or not the object is enabled and sending events to the server.
+    var isRunning: Bool { client.isRunning }
+    
+    /// Whether to show the user the analytics opt in prompt.
+    var shouldShowAnalyticsPrompt: Bool {
+        // Only show the prompt once, and when analytics are configured in BuildSettings.
+        !RiotSettings.shared.hasSeenAnalyticsPrompt && PHGPostHogConfiguration.standard != nil
+    }
+    
+    /// Indicates whether the user previously accepted Matomo analytics and should be shown the upgrade prompt.
+    var promptShouldDisplayUpgradeMessage: Bool {
+        RiotSettings.shared.hasAcceptedMatomoAnalytics
+    }
+    
+    // MARK: - Public
+    
+    /// Opts in to analytics tracking with the supplied session.
+    /// - Parameter session: An optional session to use to when reading/generating the analytics ID.
+    ///  The session will be ignored if not running.
+    func optIn(with session: MXSession?) {
+        RiotSettings.shared.enableAnalytics = true
+        startIfEnabled()
+        
+        guard let session = session else { return }
+        useAnalyticsSettings(from: session)
+    }
+    
+    /// Stops analytics tracking and calls `reset` to clear any IDs and event queues.
+    func optOut() {
+        RiotSettings.shared.enableAnalytics = false
+        
+        // The order is important here. PostHog ignores the reset if stopped.
+        reset()
+        client.stop()
+        
+        MXLog.debug("[Analytics] Stopped.")
+    }
+    
+    /// Starts the analytics client if the user has opted in, otherwise does nothing.
+    func startIfEnabled() {
+        guard RiotSettings.shared.enableAnalytics, !isRunning else { return }
+        
+        client.start()
+        
+        // Sanity check in case something went wrong.
+        guard client.isRunning else { return }
+        
+        MXLog.debug("[Analytics] Started.")
+        
+        // Catch and log crashes
+        MXLogger.logCrashes(true)
+        MXLogger.setBuildVersion(AppDelegate.theDelegate().build)
+    }
+    
+    /// Use the analytics settings from the supplied session to configure analytics.
+    /// For now this is only used for (pseudonymous) identification.
+    /// - Parameter session: The session to read analytics settings from.
+    func useAnalyticsSettings(from session: MXSession) {
+        guard
+            RiotSettings.shared.enableAnalytics,
+            !RiotSettings.shared.isIdentifiedForAnalytics
+        else { return }
+        
+        let service = AnalyticsService(session: session)
+        service.settings { result in
```

Comment:
You have to retain the `service` here, as it will be deallocated at the end of the method scope and async method will never have the chance to give a result. 
If it's currently working it's because `session` property inside the service is retained outside of it.